### PR TITLE
Feature/4997 add datadonation to settings

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Make the datadonation work from app settings screen as well
We have different behavior and texte in here

## Notice
One assert is wrong in dark mode - already requested a new version

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4997

## Screenshots
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 51 42
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 58 49
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 54 13
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 54 20
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 54 29
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 54 24
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 54 56
Simulator Screen Shot - iPhone 8 - 2021-02-12 at 15 55 06

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import Foundation
+import UIKit
+import OpenCombine
+
+/*** this protocol gets used to provide different view model classes for the dataDonationViewController */
+
+protocol DataDonationViewModelProtocol {
+
+	/// use these 3 to definde @published in a prorocol
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift:13]: 👀 don't understand this comment 🙈

**Review by tcfos — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	var reloadTableView: Bool { get }
+	var reloadTableViewPublished: OpenCombine.Published<Bool> { get }
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift:15]: Is this used somewhere?

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import Foundation
+import UIKit
+import OpenCombine
+
+/*** this protocol gets used to provide different view model classes for the dataDonationViewController */
+
+protocol DataDonationViewModelProtocol {
+
+	/// use these 3 to definde @published in a prorocol
+	var reloadTableView: Bool { get }
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift:14]: I needed some time to understand what's happening here.
I think a simple closure would be sufficient enough and a lot easier to understand.

Also a data driven approach with combine, could be dynamicTableViewModel published. Every time it changes, the tableView should be reloaded.

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	var friendlyFederalStateName: String { get }
+	var friendlyRegionName: String { get }
+	var friendlyAgeName: String { get }
+	var dynamicTableViewModel: DynamicTableViewModel { get }
+
+	func save(consentGiven: Bool)
+}
+
+class BaseDataDonationViewModel: DataDonationViewModelProtocol {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift:26]: Would be nice to replace this base implementation with a protocol extension.

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift (hunk 1 above)
Comment (marcussc) [reply src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift:13]: The whole thing is a workaround, for not being able to use property wrapper in protocols.
https://swiftsenpai.com/swift/define-protocol-with-published-property-wrapper/

Maybe the comment can be improved, which should then also explain this: https://github.com/corona-warn-app/cwa-app-ios/pull/1977#discussion_r575330743

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift (hunk 2 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift:15]: I removed the published property it's not need.

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift (hunk 1 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift]: i changed the whole thing a bit so the useless bool is gone

File: src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift (hunk 3 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/DataDonation/DataDonationViewModels/BaseDataDonationViewModel.swift]: we use the modal as a publisher right now. This bool is obsolete, I removed it.

**Pull request merged**